### PR TITLE
QR + Block Arnoldi

### PR DESCRIPTION
**Pull request by Simkern**
Here comes the updated block versions of the algorithms.

**QR algorithm**
- `qr_factorization`: Fixed unnecessary step and added comments about the second pass. 
- `test_qr_breakdown`: A special test case added to the QR test suite to check QR stability; Highlight when the single-pass GS-based QR algorithm fails.

**Block Arnoldi**
- `block_arnoldi_factorization`: Block version of the regular arnoldi factorization. Very condensed through the use of Krylov basis utilities and the QR factorization.
- `test_block_arnoldi_full_factorization` and `test_block_arnoldi_basis_orthogonality` are added to the arnoldi testsuite. They are essentially copies of the regular arnoldi tests adapted for non-unity blocks.

For now I have not touched the regular arnoldi algorithms. If we want we can either write an interface to choose the appropriate subroutine or replace the arnoldi entirely by the block version. The block size is an optional parameter for the subroutines.

### CODE REVIEW

File: src/BaseKrylov.f90 [hunk 1]
```diff
@@ -821,27 +828,157 @@ subroutine qr_factorization(Q, R, info, verbosity, tol)
...
+            enddo
+         else
+            do i = 1,p
+               call A%matvec(X(kpm+i), X(kp+i))
+            enddo
+         endif
+         ! --> Update Hessenberg matrix w.r.t. existing vectors
+         call block_update_hessenberg_matrix(H, X, k, p)
+         ! --> Orthogonalize current vectors
+         call qr_factorization(X(kp+1:kpp),H(kp+1:kpp,kpm+1:kp),info)
```
Comment (loiseaujc) [inline src/BaseKrylov.f90:903]: Why do you need to call `qr_factorization` here ?
Isn't the orthogonalization already taken care of when updating the block Hessenberg matrix ?

File: test/TestKrylov.f90 [hunk 2]
```diff
@@ -121,6 +123,99 @@ subroutine test_arnoldi_basis_orthogonality(error)
...
+      
+      ! --> Initialize matrix.
+      A = rmatrix(); call random_number(A%data)
+      ! --> Initialize Krylov subspace.
+      allocate (X(1:p*(kdim + 1))); 
+      do k = 1,p
+         call random_number(X(k)%data)
+      enddo
+      call qr_factorization(X(1:p), Rwrk, info)
+      do k = p+1, size(X)
```
Comment (loiseaujc) [inline test/TestKrylov.f90:157]: You can call `initialize_krylov_subspace` right after allocation to zero-out all the vectors before constructing the first p ones. Mostly a cosmetic change getting rid of two lines of code.

File: test/TestKrylov.f90 [hunk 3]
```diff
@@ -121,6 +123,99 @@ subroutine test_arnoldi_basis_orthogonality(error)
...
+
+      !! --> Initialize random matrix.
+      A = rmatrix(); call random_number(A%data)
+      !! --> Initialize Krylov subspace.
+      allocate (X(1:p*(kdim + 1))); 
+      do k = 1,p
+         call random_number(X(k)%data)
+      enddo
+      call qr_factorization(X(1:p), Rwrk, info)
+      do k = p+1, size(X)
```
Comment (loiseaujc) [inline test/TestKrylov.f90:200]: Same as previous comment about `initialize_krylov_subspace`.

File: test/TestKrylov.f90 [hunk 4]
```diff
@@ -121,6 +123,99 @@ subroutine test_arnoldi_basis_orthogonality(error)
...
+         call X(k)%zero()
+      enddo
+      H = 0.0_wp
+      ! --> Arnoldi factorization.
+      call block_arnoldi_factorization(A, X, H, info, block_size = p)
+      ! --> Compute Gram matrix associated to the Krylov basis.
+      do i = 1, p*kdim
```
Comment (loiseaujc) [inline test/TestKrylov.f90:207]: Can't we now change these lines with one of the `mat_something` functions you implemented ?

File: src/BaseKrylov.f90 (hunk 1 above)
Comment (Simkern) [reply src/BaseKrylov.f90:903]: It comes down to style and I took the same structure as in your arnoldi.

update hessenberg actually just fills in the rightmost column(s) of the hessenberg matrix up to and including the diagonal, i.e. orthogonalising the new vector(s) against all the previous ones. 

After this stage, the normal arnoldi would proceed to normalise the new vector and place the norm on the subdiagonal of the hessenberg matrix.
In the block arnoldi case, this normalisation is actually a QR of the p new columns.

In summary, the following parts of the code are conceptually identical.

`arnoldi_factorization`:
```
call update_hessenberg_matrix(H, X, k)
beta = X(k + 1)%norm(); H(k + 1, k) = beta
[...]
call X(k + 1)%scal(1.0_wp/beta)
```
`block_arnoldi_factorization`:
```
call block_update_hessenberg_matrix(H, X, k, p)
call qr_factorization(X(kp+1:kpp),H(kp+1:kpp,kpm+1:kp),info)
```

I understand that you separated the two steps in `arnoldi_factorization` in order to be able to do the convergence test for beta separately. You could do this in `block_arnoldi_factorization` too by creating a temporary R matrix in the QR, find the smallest beta and then add it to the Hessenberg matrix.

Should we do it this way?

File: test/TestKrylov.f90 (hunk 2 above)
Comment (Simkern) [reply test/TestKrylov.f90:157]: Correct. Also this is just an adapted copy of the code in `test_arnoldi_full_factorization`. I will change that in both routines

File: test/TestKrylov.f90 (hunk 4 above)
Comment (Simkern) [reply test/TestKrylov.f90:207]: The same is true for many other places in the code. In fact, `qr_factorization` is essentially recursively calling `update_hessenberg_matrix` for each column. Within the hessenberg/QR functions, each loop 

```
do i = 1, k
  alpha = X(k + 1)%dot(X(i)); call X(k + 1)%axpby(1.0_wp, X(i), -alpha)
  H(i, k) = alpha
end do
```
can be written as
```
call mat_mult(H(1:k,k), X(1:k), X(k+1))
call mat_mult(Xwrk(1), X(1:k), H(1:k,k))
call mat_axpby(X(k+1), 1.0_wp, Xwrk(1), -1.0_wp)
```
but I'm not sure which version is "better". The advantage is, of course, that with the right indices the code for the single step is the same for the block case making the code redundant.

Ultimately, one could refactor the code base to include these functions. I think it's more readable than the loops, but that might be down to personal taste.

File: src/BaseKrylov.f90 (hunk 1 above)
Comment (loiseaujc) [reply src/BaseKrylov.f90:903]: Ah no ok, it's fine. I did not realized it boiled down to the same thing.

**Comment by Simkern** [comment]
All the requests are implemented.

- `initalize_kylov_subspace` now takes an optional argument with the initial vector. Default is all zeros.
- `arnoldi_factorization` has been completely replaced with the block version. I have kept the separate tests for unit-size and larger blocks, it might come in handy for debugging later.
- All the examples have been refactored and streamlined, in particular regarding the initialisation.

I was thinking to replace all the orthonormalisation loops with the calls of `mat_mult` so that they would easily generalize to their block version but it turns out that this is quite messy unless the algorithms are actually written with a (possibly unit-valued) block size. This is mainly due to the fact that `mat_mult` is defined for matrix/Krylov basis-valued inputs and outputs and one would have to artificially include the extra dimension even if it is 1. 

There are 3 alternatives:

1. We simply leave the code as it is.
2. We fully generalize the algorithms to their block versions (this might require a few minor changes)
3. We generalize `mat_mult` with a specific procedure for rank 1 matrices, i.e. connect it to `get_vec` and the regular Kylov dot product.

**Review by loiseaujc — APPROVED** [review]

**Review by loiseaujc — APPROVED** [review]

**Review by loiseaujc — APPROVED** [review]

**Pull request merged**